src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,706 @@
+import type NodeGraph from './NodeGraph';
+import type { NodeId, NodeAddress, NodeData, NodeMapping } from './types';
+import type KeyManager from '../keys/KeyManager';
+import type ReverseProxy from '../network/ReverseProxy';
+import type ForwardProxy from '../network/ForwardProxy';
+import type { Host, Hostname, Port } from '../network/types';
+import type { ResourceAcquire } from '../utils';
+import type { Timer } from '../types';
+import Logger from '@matrixai/logger';
+import { StartStop, ready } from '@matrixai/async-init/dist/StartStop';
+import { IdInternal } from '@matrixai/id';
+import NodeConnection from './NodeConnection';
+import * as nodesUtils from './utils';
+import * as nodesErrors from './errors';
+import { utils as validationUtils } from '../validation';
+import * as networkUtils from '../network/utils';
+import GRPCClientAgent from '../agent/GRPCClientAgent';
+import * as agentErrors from '../agent/errors';
+import * as grpcErrors from '../grpc/errors';
+import * as nodesPB from '../proto/js/polykey/v1/nodes/nodes_pb';
+import { RWLock, withF } from '../utils';
+
+type ConnectionAndLock = {
+  connection?: NodeConnection<GRPCClientAgent>;
+  timer?: NodeJS.Timer;
+  lock: RWLock;
+};
+
+/**
+ * Data structure to store all NodeConnections. If a connection to a node n does
+ * not exist, no entry for n will exist in the map. Alternatively, if a
+ * connection is currently being instantiated by some thread, an entry will
+ * exist in the map, but only with the lock (no connection object). Once a
+ * connection is instantiated, the entry in the map is updated to include the
+ * connection object.
+ */
+type NodeConnectionMap = Map<string, ConnectionAndLock>;
+
+/**
+ * This interface is provided to each `NodeConnection` in the connection map
+ * to provide necessary features. The main one being `killSelf` to allow the
+ * `NodeConnection` to tell the `NodeConnectionManager` to destroy the
+ * connection thus removing it from  the connection map.
+ */
+export interface NodeConnectionManagerInterface {
+  withConnF: typeof NodeConnectionManager.prototype.withConnF;
+  getSeedNodes: typeof NodeConnectionManager.prototype.getSeedNodes;
+  sendHolePunchMessage: typeof NodeConnectionManager.prototype.sendHolePunchMessage;
+  killSelf(): Promise<void>;
+}
+
+interface NodeConnectionManager extends StartStop {}
+@StartStop()
+class NodeConnectionManager {
+  protected nodeGraph: NodeGraph;
+  protected keyManager: KeyManager;
+  protected fwdProxy: ForwardProxy;
+  protected revProxy: ReverseProxy;
+  protected logger: Logger;
+  // Reusing terms from network domain:
```

Comment:
These comments should be docblock comments. Not line comments. Line comments are only used inline of a function/procedure.